Implementation of Deploy/Undeploy

pcoop opened the pull request:
Implement use of a zosmf profile, addition of a cicshlq parameter, and some improvements to the async processing.

matthewpwilson commented on __tests__/cli/deploy/bundle/DeployBundle.handler.test.ts:
```diff
@@ -679,4 +679,118 @@ describe("bundle Handler", () => {
     });
+    it("should complain with no cicshlq parameter", async () => {
+
+        let parms: IHandlerParameters;
```
Suggest setting up common parms in a beforeAll function to avoid duplicating in all of these tests. You could put it inside a nested `describe` for just these tests.
In fact the whole of the test could extracted into a function which you just pass in the value of cicshlq into

pcoop replied:
I'm not sure that beforeAll would help... the parms are too different from one test to the next. But I guess some utility methods could be added where three or four tests need essentially the same values.

matthewpwilson replied:
sorry I was only referring to the new tests for --cicshlq which seem to be identical except for the value of the parameter itself, this seems like a good case to extract that test into a function and parameterise it with the value.

matthewpwilson commented on __tests__/cli/deploy/bundle/__snapshots__/DeployBundle.definition.test.ts.snap:
```diff
@@ -34,20 +35,21 @@ export const DeployBundleDefinition: ICommandDefinition = {
         {
             description: \\"Deploy a CICS bundle with a specific name and location to a default set of target regions\\",
-            options: \`--name EXAMPLE --bundledir /u/example/bundleDir --cics-deploy-profile default\`
+            options: \`--name EXAMPLE --bundledir /u/example/bundleDir --cics-deploy-profile default --zosmf-profile zprof\`
```
suggest we don't need --zosmf-profile on all of the examples as the user is likely to have a default profile

pcoop replied:
Agreed, I'll change that.

matthewpwilson commented on src/api/BundleDeploy/BundleDeployer.ts:
```diff
@@ -0,0 +1,97 @@
+   */
+  public async deployBundle(): Promise<string> {
+
+    // Validate that the parms are valid for Deploy
```
I don't think this comment adds much ;-)

matthewpwilson commented on src/api/BundleDeploy/BundleDeployer.ts:
```diff
@@ -0,0 +1,97 @@
+    }
+    catch (error) {
+      // Unit tests tend to cause an exception here
+      throw new Error("No zosmf profile found");
```
If this is just here for the unit tests then I think you could get around this by mocking `profiles.get` in the

pcoop replied:
I wanted to use an object literal function definition... but it wouldn't work. It either throws an exception at runtime, or a typescript error at compile time. I thought it might be neat to embed a template in the data that could be returned by the get() function (the tests for the sample project imply that this should be possible), but after a whole load of unhelpful error messages I got fed up with trying.

matthewpwilson replied:
Here's the code I ended up with to get this working in the test for no z/OSMF profile:
``` 
const parms = Object.assign({}, ...[DEFAULT_PARAMETERS]);

parms.arguments.name = "WIBBLE";
// .... set all the other parms.agurments ....

parms.profiles.get = () => undefined;
```
I was then able to remove the try/catch.

pcoop replied:
I've kinda adopted this, but not exactly. The problem is that removing the deep copy of DEFAULT_PARAMETERS means that every time we set a parameter, it stays set for the subsequent tests. I've restructured the tests so that we not only set the parms we want to have set, but unset those that shouldn't be set, that avoids the need for the deep copy and allows us to revert back to the approach recommended here.

matthewpwilson commented on src/cli/shared/Cicshlq.option.ts:
```diff
@@ -0,0 +1,30 @@
+    type: "string",
+    required: true,
+    stringLengthRange: [1, MAX_LENGTH],
+    description: "Specifies the High Level Qualifier (up to 35 characters) at which the CICS installation " +
```
> CICS installation datasets

Is this the right term here? Sounds to me like the datasets that you install CICS from.  Suggest 'CICS datasets'

pcoop replied:
I actually started out with "CICS datasets" but thought that was too vague! I'd hoped to distinguish between the CICS installation image and any region specific configuration that is created post installation (trace file, csd, etc). I looked for standard terminology around hlqs, and found several different ways of describing the concept, none of which seemed especially good. Anyway, happy to switch it back.

pcoop commented:
I've added mocking to the unit tests and refactored accordingly. I suggest we promote the question of whether to auto-undeploy as part of deploy to an issue for wider discussion.
